walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,19 +15,137 @@
+   *
+   * @return: a sorted list of available meeting times in chronological order
+   */
+  private ArrayList<TimeRange> findAvailableMeetingTimes(Collection<Event> events, MeetingRequest request, boolean hasOptionalAttendees){
+
+    // Creates a list of all the time ranges in a day to check for conflicts with {@code events}
+    ArrayList<TimeRange> availableMeetingTimes = new ArrayList<TimeRange>();
+    availableMeetingTimes.add(TimeRange.WHOLE_DAY);
 
+    // Iterates through all the scheduled events to find any conflicts
+    for(Event event : events){
+      TimeRange eventTime = event.getWhen();
+      
+      /* Checks if the {@code event} has any attendees that are also an attendee for the new meeting {@code request}
+       * If the two collections are disjoint, skip this event
+       */
+      if(checkIfEventsAreDisjoint(event, request, hasOptionalAttendees)){
+          continue;
+      }
+      // List of all times that the meeting can take place
+      ArrayList<TimeRange> addTimes = new ArrayList<TimeRange>();
+      // List of all the times the meeting cannot take place
+      ArrayList<TimeRange> removeTimes = new ArrayList<TimeRange>();
+
+      // Checks if the {@code eventTime} overlaps with the current {@code time}
+      for(TimeRange time: availableMeetingTimes){
+          if(time.overlaps(eventTime)) {
+            timesAroundEvent(addTimes, removeTimes, request.getDuration(), time, eventTime);
+          }
+      }
+      availableMeetingTimes.removeAll(removeTimes);
+      availableMeetingTimes.addAll(addTimes);
+      }
+      return sortArray(availableMeetingTimes);
+    }
+  
+  /*
+   * Resolves conflicting event times by checking before and after the {@code eventTime} for
+   * available times with appropriate {@code newMeetingDuration}
+   *
+   * @param addTimes: list of all the available time ranges for the new meeting
+   * @param removeTimes: list of all the time ranges where the meeting cannot be held
+   * @param newMeetingDuration: the duration of requested meeting
+   * @param time: The conflicting time that the function is checking around
+   * @param eventTime: the time of the event 
+   */ 
+  private void timesAroundEvent(ArrayList<TimeRange> addTimes, ArrayList<TimeRange> removeTimes, long newMeetingDuration, TimeRange time, TimeRange eventTime) {
+    // Create a Time Range for before the {@code eventTime}
+    TimeRange beforeConflict = TimeRange.fromStartEnd(time.start(), eventTime.start(), false);
+    // Create a Time Range for after teh {@code eventTime}
+    TimeRange afterConflict = TimeRange.fromStartEnd(eventTime.end(), time.end(), false);
+
+    // Checks if the new Time Ranges created are longer than the {@code newMeetingDuration}
+    if(beforeConflict.duration() >= newMeetingDuration){
+      addTimes.add(beforeConflict);
+    }
+    if(afterConflict.duration() >= newMeetingDuration){
+      addTimes.add(afterConflict);
+    }
+    removeTimes.add(time);
```

Comment:
Since this method is always telling the caller to remove `time` from its list, I think the caller should just do that automatically. Once `removeTimes` is no longer used, this method could just return the `TimeRanges` to add as its return value, which is easier to follow than having two parameters that are used for output.

Sometimes, especially when there are multiple pieces of output data, it is necessary or more intuitive to use parameters for output, but in general, providing output via return value is easier to follow.

Reply:
Update: I forgot that this is being used while iterating over the List that the element would be deleted from, so it's not quite as simple as having the caller delete the item directly, but I still think that this method should not be concerned with the deletion. I think that the caller should still maintain a list of times to delete, since it can't immediately delete them while iterating through them*. Just that the list is built inside that method itself.

* Well it could if using an `Iterator` explicitly, but I don't think it's worth adding a little bit of complexity to do so.